Arduino Compilation Broke

using git bisect the issue seems to have been introduced here c0c6681c22186015861f1f3445f584a49769aea2
 
```
In file included from c:\program files (x86)\arduino\hardware\tools\arm\arm-none-eabi\include\c++\4.8.4\bits\stl_algo.h:60:0,

                 from c:\program files (x86)\arduino\hardware\tools\arm\arm-none-eabi\include\c++\4.8.4\algorithm:62,

                 from C:\Users\benhu\AppData\Local\Temp\arduino_build_230347\sketch\Maze.cpp:4:

c:\program files (x86)\arduino\hardware\tools\arm\arm-none-eabi\include\c++\4.8.4\bits\algorithmfwd.h:360:41: error: macro "max" passed 3 arguments, but takes just 2

     max(const _Tp&, const _Tp&, _Compare);

                                         ^

c:\program files (x86)\arduino\hardware\tools\arm\arm-none-eabi\include\c++\4.8.4\bits\algorithmfwd.h:371:41: error: macro "min" passed 3 arguments, but takes just 2

     min(const _Tp&, const _Tp&, _Compare);

                                         ^

c:\program files (x86)\arduino\hardware\tools\arm\arm-none-eabi\include\c++\4.8.4\bits\algorithmfwd.h:394:30: error: macro "min" requires 2 arguments, but only 1 given

     min(initializer_list<_Tp>);

                              ^
```

--- a/Micromouse/ArduinoSpoof/Arduino.cpp
+++ b/Micromouse/ArduinoSpoof/Arduino.cpp
@@ -153,7 +153,7 @@ void mock::Teensy::analogWrite(uint8_t pin, int val) {
 uint8_t mock::Teensy::digitalRead(uint8_t pin) {
     Pin &p = pins[pin];
 
-    if (p.mode != INPUT || p.mode != INPUT_PULLUP || p.mode != INPUT_PULLDOWN) {
+    if (p.mode != INPUT && p.mode != INPUT_PULLUP && p.mode != INPUT_PULLDOWN) {
         Serial.printf("Waring: Pin %u is not set to INPUT\n", pin);
     }
 
@@ -163,7 +163,7 @@ uint8_t mock::Teensy::digitalRead(uint8_t pin) {
 int mock::Teensy::analogRead(uint8_t pin) {
     Pin &p = pins[pin];
 
-    if (p.mode != INPUT || p.mode != INPUT_PULLUP || p.mode != INPUT_PULLDOWN) {
+    if (p.mode != INPUT && p.mode != INPUT_PULLUP && p.mode != INPUT_PULLDOWN) {
         Serial.printf("Waring: Pin %u is not set to INPUT\n", pin);
     }
 

--- a/Micromouse/Hardware.cpp
+++ b/Micromouse/Hardware.cpp
@@ -11,7 +11,11 @@ const unsigned Hardware::COUNT_PER_NODE =
     Hardware::COUNT_PER_MM * Hardware::MM_PER_NODE;
 
 Hardware::Hardware()
-    : led(Hardware::LED_PIN)
+    : leftMotor(MOTOR_LEFT_EN_PIN, MOTOR_LEFT_IN1_PIN, MOTOR_LEFT_IN2_PIN,
+                ENCODER_LEFT_A_PIN, ENCODER_LEFT_B_PIN)
+    , rightMotor(MOTOR_RIGHT_EN_PIN, MOTOR_RIGHT_IN1_PIN, MOTOR_RIGHT_IN2_PIN,
+                 ENCODER_RIGHT_A_PIN, ENCODER_RIGHT_B_PIN)
+    , led(LED_PIN)
     , speedPID(1.0f, 1.0f, 1.0f, 0.0f, 100.0f)
     , distancePID(1.0f, 1.0f, 1.0f, 0.0f, 100.0f)
     , leftPID(1.0f, 1.0f, 1.0f, 0.0f, 100.0f)

--- a/Micromouse/Hardware.h
+++ b/Micromouse/Hardware.h
@@ -1,10 +1,13 @@
 #pragma once
+#include <Arduino.h>
+#undef min
+#undef max
+
 #include "Direction.h"
 #include "Led.h"
 #include "Motor.h"
 #include "PIDController.h"
 #include "RangeFinder.h"
-#include <Arduino.h>
 
 class Hardware {
   public:

--- a/Micromouse/Hardware.h
+++ b/Micromouse/Hardware.h
@@ -11,14 +11,31 @@ class Hardware {
     // PIN DEFINITIONS
 
     // TODO: set correct pin values
-    static const unsigned IRSENSOR_LEFT_PIN = 0;
-    static const unsigned IRSENSOR_RIGHT_PIN = 0;
-    static const unsigned ULTRASONIC_PIN = 0;
-    // static const unsigned MOTOR_LEFT_PIN = 17
-    // ...
-    // ...
-    // static const unsigned RANGE_FINDER_FRONT_PIN = 18
+    static const unsigned IRSENSOR_LEFT_PIN = 16;
+    static const unsigned IRSENSOR_RIGHT_PIN = 17;
+    static const unsigned ULTRASONIC_PIN = 14;
+
+    static const unsigned MOTOR_LEFT_EN_PIN = 4;
+    static const unsigned MOTOR_LEFT_IN1_PIN = 8;
+    static const unsigned MOTOR_LEFT_IN2_PIN = 7;
+    static const unsigned ENCODER_LEFT_A_PIN = 21;
+    static const unsigned ENCODER_LEFT_B_PIN = 20;
+
+    static const unsigned MOTOR_RIGHT_EN_PIN = 3;
+    static const unsigned MOTOR_RIGHT_IN1_PIN = 6;
+    static const unsigned MOTOR_RIGHT_IN2_PIN = 5;
+    static const unsigned ENCODER_RIGHT_A_PIN = 23;
+    static const unsigned ENCODER_RIGHT_B_PIN = 22;
+
+    static const unsigned DIP_1_PIN = 35;
+    static const unsigned DIP_2_PIN = 36;
+    static const unsigned DIP_3_PIN = 37;
+    static const unsigned DIP_4_PIN = 38;
+
+    static const unsigned BUTTON_PIN = 38;
+
     static const unsigned LED_PIN = 13;
+    // static const unsigned RGB_LED_PIN = 2; // TODO update interface
 
     static const unsigned NUM_RANGE_FINDERS = 3;
     static const unsigned IRSENSOR_LEFT_MIN_DISTANCE = 20;   // mm

--- a/Micromouse/IRSensor.cpp
+++ b/Micromouse/IRSensor.cpp
@@ -1,5 +1,8 @@
-#include "IRSensor.h"
 #include <Arduino.h>
+#undef min
+#undef max
+
+#include "IRSensor.h"
 #include <cmath>
 
 float IRSensor::getDistance() const {

--- a/Micromouse/Led.cpp
+++ b/Micromouse/Led.cpp
@@ -1,5 +1,8 @@
-#include "Led.h"
 #include <Arduino.h>
+#undef min
+#undef max
+
+#include "Led.h"
 
 Led::Led() : PIN(0) {
 }

--- a/Micromouse/Maze.cpp
+++ b/Micromouse/Maze.cpp
@@ -200,8 +200,10 @@ Maze Maze::generate(uint16_t seed) {
         // maze.reset();
     }
 
+#if !defined(__MK66FX1M0__) && !defined(__MK20DX256__)
     float currentSimulationSpeed = SIMULATION_SPEED;
     SIMULATION_SPEED = numeric_limits<float>::infinity();
+#endif
 
     int removalAttempts = 30;
     // 30 + 3d20 expected value = 61.5
@@ -247,7 +249,9 @@ Maze Maze::generate(uint16_t seed) {
         //              removal.y, maze.path.size());
     }
 
+#if !defined(__MK66FX1M0__) && !defined(__MK20DX256__)
     SIMULATION_SPEED = currentSimulationSpeed;
+#endif
 
     return maze;
 }

--- a/Micromouse/Maze.cpp
+++ b/Micromouse/Maze.cpp
@@ -1,5 +1,8 @@
-#include "Maze.h"
 #include <Arduino.h>
+#undef min
+#undef max
+
+#include "Maze.h"
 #include <algorithm>
 #include <assert.h>
 #include <cstdlib> //abs

--- a/Micromouse/Micromouse.ino
+++ b/Micromouse/Micromouse.ino
@@ -10,7 +10,7 @@ void __throw_length_error(char const *e) {
     Serial.printf("Length Error : %s\n", e);
     exit(1);
 }
-}
+} // namespace std
 
 extern "C" {
 int _getpid() {
@@ -30,7 +30,12 @@ void setup() {
     Serial.begin(9600);
     delay(2000);
     Mouse mouse;
-    // mouse.testMode(Mouse::TestMode::TEST_MOTOR_SINGLE);
+    //mouse.testMode(Mouse::TestMode::TEST_MOTOR_SINGLE);
+    //mouse.testMode(Mouse::TestMode::TEST_MOTOR_PAIR);
+    //mouse.testMode(Mouse::TestMode::TEST_MOVEMENT);
+    //mouse.testMode(Mouse::TestMode::TEST_ROTATE);
+    //mouse.testMode(Mouse::TestMode::TEST_RANGE_FINDER_SINGLE);
+    //mouse.testMode(Mouse::TestMode::TEST_RANGE_FINDER_ALL);
 }
 
 void loop() {

--- a/Micromouse/Motor.cpp
+++ b/Micromouse/Motor.cpp
@@ -1,5 +1,8 @@
-#include "Motor.h"
 #include <Arduino.h>
+#undef min
+#undef max
+
+#include "Motor.h"
 #include <cmath>
 
 Motor::Motor(void) : Motor(0, 0, 0, 0, 0) {

--- a/Micromouse/Motor.cpp
+++ b/Micromouse/Motor.cpp
@@ -5,17 +5,20 @@
 Motor::Motor(void) : Motor(0, 0, 0, 0, 0) {
 }
 
-Motor::Motor(unsigned en, unsigned in1, unsigned in2, unsigned enc1,
-             unsigned enc2)
+Motor::Motor(unsigned en, unsigned in1, unsigned in2, unsigned encA,
+             unsigned encB)
     : en(en)
     , in1(in1)
     , in2(in2)
 #if defined(__MK66FX1M0__) || defined(__MK20DX256__)
-    , encoder(enc1, enc2)
-#endif
+    , encoder(encA, encB) {
+#else
 {
-    (void)enc1;
-    (void)enc2;
+    // Just so the pins are passed to our mock Teensy
+    pinMode(encA, INPUT);
+    pinMode(encB, INPUT);
+#endif
+
     pinMode(en, OUTPUT);
     pinMode(in1, OUTPUT);
     pinMode(in2, OUTPUT);

--- a/Micromouse/Motor.h
+++ b/Micromouse/Motor.h
@@ -9,8 +9,8 @@
 class Motor {
   public:
     Motor(void);
-    Motor(unsigned en, unsigned in1, unsigned in2, unsigned enc1,
-          unsigned enc2);
+    Motor(unsigned en, unsigned in1, unsigned in2, unsigned encA,
+          unsigned encB);
     ~Motor();
 
     void off(void);

--- a/Micromouse/Mouse.cpp
+++ b/Micromouse/Mouse.cpp
@@ -418,10 +418,13 @@ void Mouse::mapMazeDFS() {
         maze.setExplored(Maze::NODE_START);
         maze.setExplored(Maze::NODE_FINISH);
 
+// TODO replace with hardware function that uses sensors
+#if !defined(__MK66FX1M0__) && !defined(__MK20DX256__)
         for (auto relation : RELATIONS) {
             if (virtualMaze.isWall(position, DirOp::relToDir(relation, facing)))
                 maze.setWall(position, DirOp::relToDir(relation, facing));
         }
+#endif
 
         maze.setExplored(position);
         maze.getNode(position).visited = true;

--- a/Micromouse/Mouse.cpp
+++ b/Micromouse/Mouse.cpp
@@ -545,11 +545,11 @@ void Mouse::rankMappingStrategies(const unsigned cycles) {
     }
 
     for (auto strategy : MAP_STRATS) {
-        Serial.printf("S:%i, Moves:(%4u, %4u, %4u) Costs:(%6u, %6u, %6u) \n", strategy,
-                      stats[strategy].min.moves, stats[strategy].max.moves,
-                      stats[strategy].avg.moves / cycles,
-                      stats[strategy].min.cost, stats[strategy].max.cost,
-                      stats[strategy].avg.cost / cycles);
+        Serial.printf(
+            "S:%i, Moves:(%4u, %4u, %4u) Costs:(%6u, %6u, %6u) \n", strategy,
+            stats[strategy].min.moves, stats[strategy].max.moves,
+            stats[strategy].avg.moves / cycles, stats[strategy].min.cost,
+            stats[strategy].max.cost, stats[strategy].avg.cost / cycles);
     }
 }
 #endif

--- a/Micromouse/Mouse.cpp
+++ b/Micromouse/Mouse.cpp
@@ -1,5 +1,8 @@
-#include "Mouse.h"
 #include <Arduino.h>
+#undef min
+#undef max
+
+#include "Mouse.h"
 #include <assert.h>
 #include <cstdio>
 #include <stack>

--- a/Micromouse/RangeFinder.cpp
+++ b/Micromouse/RangeFinder.cpp
@@ -1,5 +1,8 @@
-#include "RangeFinder.h"
 #include <Arduino.h>
+#undef min
+#undef max
+
+#include "RangeFinder.h"
 #include <assert.h>
 
 RangeFinder::RangeFinder() : DATA_PIN(0), MIN_RANGE(0), MAX_RANGE(0) {

--- a/Simulation/simulate.cpp
+++ b/Simulation/simulate.cpp
@@ -359,6 +359,14 @@ class Simulator : public sf::RenderWindow {
 int main() {
     srand(time(0));
     Mouse mouse;
+
+    // mouse.testMode(Mouse::TestMode::TEST_MOTOR_SINGLE);
+    // mouse.testMode(Mouse::TestMode::TEST_MOTOR_PAIR);
+    // mouse.testMode(Mouse::TestMode::TEST_MOVEMENT);
+    // mouse.testMode(Mouse::TestMode::TEST_ROTATE);
+    // mouse.testMode(Mouse::TestMode::TEST_RANGE_FINDER_SINGLE);
+    // mouse.testMode(Mouse::TestMode::TEST_RANGE_FINDER_ALL);
+
     Simulator simulator(mouse);
     simulator.main_loop();
     return 0;

